File: programs/protocol/src/data/offer.rs
use anchor_lang::prelude::*;

#[zero_copy]
#[derive(Debug, Default, PartialEq, Eq)]
pub struct OfferDetails {
    pub payment: u64,
    pub collateral: u64,
    pub deadline: u64,
}

#[account(zero_copy)]
#[derive(Debug, Default, PartialEq)]
pub struct ShipmentOffer {
    pub offeror: Pubkey,
    pub shipment: Pubkey,
    pub details: OfferDetails,
    pub submitted: i64,
    pub timeout: i64,
    pub no: u32,
    pub reserved: [u8; 4],
}

#[account(zero_copy)]
#[derive(Debug, Default, PartialEq)]
pub struct AcceptedOffer {
    pub owner: Pubkey,
    pub shipment: Pubkey,
    pub details: OfferDetails,
    pub accepted: i64,
    pub no: u32,
    pub reserved: [u8; 4],
}

File: app/src/utils/parse/offer.ts
import type BN from 'bn.js';
import type { ShipmentOffer, ApiShipmentOffer, OfferDetails } from '../account/offer';


export function parseOfferDetails(details: OfferDetails<BN, BN>): OfferDetails<number, string> {
	return {
		payment: details.payment.toNumber(),
		collateral: details.collateral.toNumber(),
		deadline: details.deadline.toString()
	};
}

export function parseOfferToApiOffer(offerAccount: ShipmentOffer): ApiShipmentOffer {
	return {
		...offerAccount,
		offeror: offerAccount.offeror.toString(),
		submitted: new Date(offerAccount.submitted.toNumber() * 1000).toISOString(),
		timeout: new Date(offerAccount.timeout.toNumber() * 1000).toISOString(),
		shipment: offerAccount.shipment.toString(),
		details: parseOfferDetails(offerAccount.details)
	};
}

